Naming requirements for annotated types depend on Python version (pylint 4 regression)

### Bug description

pylint requires different naming conventions for annotated types (`typing.Annotated`) depending on the Python version used:

- under 3.11 it requires PascalCase
- under 3.13 it requires snake_case

I’m using a simple example file (`annotated.py`) to reproduce this:

```python
"""Test-case for annotations."""

from typing import Annotated

int_with_annotation = Annotated[int, "anything"]  # Error in Python 3.11
IntWithAnnotation = Annotated[int, "anything"]    # Error in Python 3.13
```

### Configuration

```ini

```

### Command used

```shell
pylint annotated.py
```

### Pylint output

```
Python 3.11
************* Module annotated
annotated.py:5:0: C0103: Class name "int_with_annotation" doesn't conform to PascalCase naming style (invalid-name)

Python 3.13
************* Module annotated
annotated.py:5:0: C0103: Variable name "IntWithAnnotation" doesn't conform to snake_case naming style (invalid-name)
```

### Expected behavior

I would expect pylint to have the same opinion on naming conventions regardless of the (supported) Python version in use.

### Pylint version

```
pylint 4.0.2
astroid 4.0.1
Python 3.11.13 (main, Jul 17 2025, 10:22:15) [Clang 20.1.7 ]

---

pylint 4.0.2
astroid 4.0.1
Python 3.13.5 (main, Jul 17 2025, 10:33:11) [Clang 20.1.7 ]
```

### OS / Environment

_No response_

### Additional dependencies

```python

```

--- a/doc/whatsnew/fragments/10696.false_positive
+++ b/doc/whatsnew/fragments/10696.false_positive
@@ -1,0 +1,3 @@
+Fixed false positive for ``invalid-name`` with ``typing.Annotated``.
+
+Closes #10696

--- a/pylint/checkers/base/name_checker/checker.py
+++ b/pylint/checkers/base/name_checker/checker.py
@@ -488,17 +488,7 @@ class NameChecker(_BasicChecker):
                     return
 
                 # Check classes (TypeVar's are classes so they need to be excluded first)
-                elif isinstance(inferred_assign_type, nodes.ClassDef) or (
-                    isinstance(inferred_assign_type, bases.Instance)
-                    and {"EnumMeta", "TypedDict"}.intersection(
-                        {
-                            ancestor.name
-                            for ancestor in cast(
-                                InferenceResult, inferred_assign_type
-                            ).mro()
-                        }
-                    )
-                ):
+                elif self._should_check_class_regex(inferred_assign_type):
                     self._check_name("class", node.name, node)
 
                 # Don't emit if the name redefines an import in an ImportError except handler
@@ -575,6 +565,28 @@ class NameChecker(_BasicChecker):
             return False
         regexp = self._name_regexps["variable"]
         return regexp.match(name) is not None
+
+    def _should_check_class_regex(
+        self, inferred_assign_type: InferenceResult | None
+    ) -> bool:
+        if isinstance(inferred_assign_type, nodes.ClassDef):
+            return True
+        if isinstance(inferred_assign_type, bases.Instance) and {
+            "EnumMeta",
+            "TypedDict",
+        }.intersection(
+            {
+                ancestor.name
+                for ancestor in cast(InferenceResult, inferred_assign_type).mro()
+            }
+        ):
+            return True
+        if (
+            isinstance(inferred_assign_type, nodes.FunctionDef)
+            and inferred_assign_type.qname() == "typing.Annotated"
+        ):
+            return True
+        return False
 
     def _recursive_check_names(self, args: list[nodes.AssignName]) -> None:
         """Check names in a possibly recursive list <arg>."""

--- a/tests/functional/i/invalid/invalid_name/invalid_name_module_level.py
+++ b/tests/functional/i/invalid/invalid_name/invalid_name_module_level.py
@@ -51,3 +51,7 @@ try:
     VERSION = version("ty")  # uninferable
 except PackageNotFoundError:
     VERSION = "0.0.0"
+
+
+from typing import Annotated
+IntWithAnnotation = Annotated[int, "anything"]

